core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/pillar/DefaultPillarSelector.java

```diff
@@ -111,44 +93,14 @@ public SelectionCacheType getCacheType() {
 
     @Override
     public void constructCache(SolverScope<Solution_> solverScope) {
-        long entitySize = entitySelector.getSize();
-        if (entitySize > Integer.MAX_VALUE) {
-            throw new IllegalStateException("The selector (" + this + ") has an entitySelector ("
-                    + entitySelector + ") with entitySize (" + entitySize
-                    + ") which is higher than Integer.MAX_VALUE.");
-        }
-        Stream<Object> entities = StreamSupport.stream(entitySelector.spliterator(), false);
-        Comparator<?> comparator = subpillarConfigPolicy.getEntityComparator();
-        if (comparator != null) {
-            /*
-             * The entity selection will be sorted. This will result in all the pillars being sorted without having to
-             * sort them individually later.
-             */
-            entities = entities.sorted((Comparator<? super Object>) comparator);
-        }
-        // Create all the pillars from a stream of entities; if sorted, the pillars will be sequential.
-        Map<List<Object>, List<Object>> valueStateToPillarMap = new LinkedHashMap<>((int) entitySize);
-        int variableCount = variableDescriptors.size();
-        entities.forEach(entity -> {
-            List<Object> valueState = variableCount == 1 ? getSingleVariableValueState(entity, variableDescriptors)
-                    : getMultiVariableValueState(entity, variableDescriptors, variableCount);
-            List<Object> pillar = valueStateToPillarMap.computeIfAbsent(valueState, key -> new ArrayList<>());
-            pillar.add(entity);
-        });
-        // Store the cache. Exclude pillars of size lower than the minimumSubPillarSize, as we shouldn't select those.
-        Collection<List<Object>> pillarLists = valueStateToPillarMap.values();
-        int minimumSubPillarSize = subpillarConfigPolicy.getMinimumSubPillarSize();
-        if (minimumSubPillarSize > 1) {
-            cachedBasePillarList = pillarLists.stream()
-                    .filter(pillar -> pillar.size() >= minimumSubPillarSize)
-                    .collect(Collectors.toList());
-        } else { // Use shortcut when we don't intend to remove anything.
-            cachedBasePillarList = new ArrayList<>(pillarLists);
-        }
+        cachedBasePillarList = solverScope.getPillarSupplyManager()
+                .demand(pillarDemand)
+                .getPillars();
     }
 
     @Override
     public void disposeCache(SolverScope<Solution_> solverScope) {
+        ((PillarSupplyManager) solverScope.getPillarSupplyManager()).reset();
         cachedBasePillarList = null;
```

Comment:
The cache is short-lived, possibly only step-scoped.
But that is still better than no cache.